dht/token.hh

```diff
@@ -213,7 +213,25 @@ struct raw_token_less_comparator {
 
 const token& minimum_token() noexcept;
 const token& maximum_token() noexcept;
-std::strong_ordering operator<=>(const token& t1, const token& t2);
+
+inline int64_t long_token(const token& t) {
+    if (t.is_minimum() || t.is_maximum()) {
+        return std::numeric_limits<int64_t>::min();
+    }
+
+    return t._data;
+}
+
+inline std::strong_ordering operator<=>(const token& t1, const token& t2) {
+    if (t1._kind < t2._kind) {
+        return std::strong_ordering::less;
```

Comment:
We should remove token::_kind one day.

Reply:
Yeah.  @michoecho started some work in this direction.  Curreny we have too many ways to represent ranges.  I think that before/after all tokens concept belongs to bounds.
And then there are intervals with missing bounds that also stand for infinity (a.k.a unbound/semi-bound ranges)

Reply:
Maybe we should simply keep tokens as signed 128 bit numbers and then the kind would be essentially in the most significant word.

Reply:
> Yeah. @michoecho started some work in this direction.

More than just started. Removing `token::_kind` was indeed one of the things done by https://github.com/scylladb/scylladb/pull/8806, more than 3 years ago.

That PR was basically finished, but at the time I was too much of a noob and too scared of breaking something (especially in the thrift API) to get it merged.

Reply:
> Maybe we should simply keep tokens as signed 128 bit numbers and then the kind would be essentially in the most significant word.

This idea might have merit...
See https://godbolt.org/z/5jWv5bsef

Reply:
> > Yeah. @michoecho started some work in this direction.
> 
> More than just started. Removing `token::_kind` was indeed one of the things done by #8806, more than 3 years ago.
> 
> That PR was basically finished, but at the time I was too much of a noob and too scared of breaking something (especially in the thrift API) to get it merged.

Now that you're experienced you should be even more scared.

I think the hard parts are when crossing node boundaries, since you have to work with old nodes. But I expect all tokens there will be part of intervals, so we can introduce helpers to convert interval\<heavy_token> to interval\<token>.